File: python/Ans41.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        import random
        def findMedian(nums, start, end):
            if end <= start:
                return
            pivotIndex = random.randrange(start, end)
            nums[start], nums[pivotIndex] = nums[pivotIndex], nums[start]
            i = start + 1
            j = end
            while i <= j:
                if nums[i] > nums[start]:
                    nums[i], nums[j] = nums[j], nums[i]
                    j -= 1
                else:
                    i += 1
            nums[start], nums[j] = nums[j], nums[start]
            findMedian(nums, j + 1, end)
            findMedian(nums, start, j - 1)
        
        findMedian(nums, 0, len(nums) - 1)
        num1 = 1
        for i in range(len(nums)):
            if num1 == nums[i]:
                num1 += 1
        return num1

File: java/Ans41.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        int i=0;
        while(i<nums.length){
            int correct=nums[i]-1;
            if(nums[i]<nums.length&&nums[i]>0 && nums[i]!=nums[correct]){
                swap(nums,i,correct);
            }else{
                i++;
            }
        }
        int counter=0;
        
        for(int j=0;j<nums.length;j++){
            if(nums[j]!=j+1){
                return j+1;
            }else{
                counter++;
            }  
    
        }

        if(counter==nums.length)
            return counter+1;

        return -1;
    }

    public void swap(int[] arr, int a, int b){
        int temp=arr[a];
        arr[a]=arr[b];
        arr[b]=temp;
    }
}
